### PR TITLE
Add python 3.10 support

### LINKED ISSUE #2607: Hitting  `Fields of type "<class 'types.Union'>" are not supported` in Python 3.10.0a6 when using the ' | ' for Union in annotating field types.
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug
I'm hitting  `Fields of type "<class 'types.Union'>" are not supported` in Python 3.10.0a6 when using the ' | ' for Union in annotating field types. However, when I use the `Union` annotation for field types, it works:

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.1
            pydantic compiled: False
                 install path: /home/donato/.pyenv/versions/3.10.0a6/envs/py3_10a6/lib/python3.10/site-packages/pydantic
               python version: 3.10.0a6 (default, Mar 30 2021, 12:26:09) [GCC 9.3.0]
                     platform: Linux-5.4.72-microsoft-standard-WSL2-x86_64-with-glibc2.31
     optional deps. installed: ['typing-extensions']
```

```
Python 3.10.0a6 (default, Mar 30 2021, 12:26:09) [GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pydantic
>>> from typing import Union
>>> class TestB(pydantic.BaseModel):
...   f1: Union[str, None]
...   f2: Union[int, None]
...
>>> t1 = TestB()
>>> t1
TestB(f1=None, f2=None)
>>> class TestC(pydantic.BaseModel):
...   f1: str | None
...   f2: int | None
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/donato/.pyenv/versions/py3_10a6/lib/python3.10/site-packages/pydantic/main.py", line 293, in __new__
    fields[ann_name] = ModelField.infer(
  File "/home/donato/.pyenv/versions/py3_10a6/lib/python3.10/site-packages/pydantic/fields.py", line 410, in infer
    return cls(
  File "/home/donato/.pyenv/versions/py3_10a6/lib/python3.10/site-packages/pydantic/fields.py", line 342, in __init__
    self.prepare()
  File "/home/donato/.pyenv/versions/py3_10a6/lib/python3.10/site-packages/pydantic/fields.py", line 450, in prepare
    self._type_analysis()
  File "/home/donato/.pyenv/versions/py3_10a6/lib/python3.10/site-packages/pydantic/fields.py", line 626, in _type_analysis
    raise TypeError(f'Fields of type "{origin}" are not supported.')
TypeError: Fields of type "<class 'types.Union'>" are not supported.
>>>
```

### LINKED ISSUE #2786: Deprecation warning emitted by "generate_model_signature"
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

With Python 3.10.0b1 a deprecation warning is emitted from the `pydantic.utils.generate_model_signature` function.
This is due to the following line:

https://github.com/samuelcolvin/pydantic/blob/b718e8e62680c2d7a42f1cf946094e1c0fe28c1d/pydantic/utils.py#L234

Accessing one enum member from another member is discouraged and is deprecated in Python 3.10, with removal scheduled for 3.12. See [bpo-43162](https://bugs.python.org/issue43162) and https://github.com/python/cpython/pull/24486 for more information.

This can be reproduced as follows:

`python3.10 -Werror -c "import pydantic"`

<details><summary>Traceback</summary>

```python traceback
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.10/site-packages/pydantic/__init__.py", line 2, in <module>
    from . import dataclasses
  File "/usr/lib/python3.10/site-packages/pydantic/dataclasses.py", line 7, in <module>
    from .main import create_model, validate_model
  File "/usr/lib/python3.10/site-packages/pydantic/main.py", line 376, in <module>
    class BaseModel(Representation, metaclass=ModelMetaclass):
  File "/usr/lib/python3.10/site-packages/pydantic/main.py", line 369, in __new__
    cls.__signature__ = ClassAttribute('__signature__', generate_model_signature(cls.__init__, fields, config))
  File "/usr/lib/python3.10/site-packages/pydantic/utils.py", line 235, in generate_model_signature
    if config.extra is config.extra.allow:
  File "/usr/lib/python3.10/enum.py", line 151, in __get__
    warnings.warn(
DeprecationWarning: accessing one member from another is not supported,  and will be disabled in 3.12
```

</details>

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.2
            pydantic compiled: False
                 install path: /usr/lib/python3.10/site-packages/pydantic
               python version: 3.10.0b1 (default, May  6 2021, 02:43:13) [GCC 9.3.0]
                     platform: Linux-5.8.0-50-generic-x86_64-with-glibc2.31
     optional deps. installed: ['typing-extensions']

```

### LINKED ISSUE #2787: Python 3.10 test failures due to change in enum repr
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

This is related to https://bugs.python.org/issue40066 that changed the format of repr output.

```
_______________________________________________________________________ test_literal_enum_values _______________________________________________________________________

    def test_literal_enum_values():
        FooEnum = Enum('FooEnum', {'foo': 'foo_value', 'bar': 'bar_value'})
    
        class Model(BaseModel):
            baz: Literal[FooEnum.foo]
            boo: str = 'hoo'
    
            class Config:
                use_enum_values = True
    
        m = Model(baz=FooEnum.foo)
        assert m.dict() == {'baz': 'foo_value', 'boo': 'hoo'}
        assert m.baz.value == 'foo_value'
    
        with pytest.raises(ValidationError) as exc_info:
            Model(baz=FooEnum.bar)
    
>       assert exc_info.value.errors() == [
            {
                'loc': ('baz',),
                'msg': "unexpected value; permitted: <FooEnum.foo: 'foo_value'>",
                'type': 'value_error.const',
                'ctx': {'given': FooEnum.bar, 'permitted': (FooEnum.foo,)},
            },
        ]
E       assert [{'ctx': {'given': FooEnum.bar, 'permitted': (FooEnum.foo,)},\n  'loc': ('baz',),\n  'msg': 'unexpected value; permitted: FooEnum.foo',\n  'type': 'value_error.const'}] == [{'ctx': {'given': FooEnum.bar, 'permitted': (FooEnum.foo,)},\n  'loc': ('baz',),\n  'msg': "unexpected value; permitted: <FooEnum.foo: 'foo_value'>",\n  'type': 'value_error.const'}]
E         At index 0 diff: {'loc': ('baz',), 'msg': 'unexpected value; permitted: FooEnum.foo', 'type': 'value_error.const', 'ctx': {'given': FooEnum.bar, 'permitted': (FooEnum.foo,)}} != {'loc': ('baz',), 'msg': "unexpected value; permitted: <FooEnum.foo: 'foo_value'>", 'type': 'value_error.const', 'ctx': {'given': FooEnum.bar, 'permitted': (FooEnum.foo,)}}
E         Full diff:
E           [
E            {'ctx': {'given': FooEnum.bar, 'permitted': (FooEnum.foo,)},
E             'loc': ('baz',),
E         -   'msg': "unexpected value; permitted: <FooEnum.foo: 'foo_value'>",
E         ?          ^                             -           -- ------------
E         +   'msg': 'unexpected value; permitted: FooEnum.foo',
E         ?          ^
E             'type': 'value_error.const'},
E           ]

tests/test_main.py:806: AssertionError
_________________________________________________________________________ test_enum_successful _________________________________________________________________________

    def test_enum_successful():
        m = CookingModel(tool=2)
        assert m.fruit == FruitEnum.pear
        assert m.tool == ToolEnum.wrench
>       assert repr(m.tool) == '<ToolEnum.wrench: 2>'
E       AssertionError: assert 'ToolEnum.wrench' == '<ToolEnum.wrench: 2>'
E         - <ToolEnum.wrench: 2>
E         ? -               ----
E         + ToolEnum.wrench

tests/test_types.py:836: AssertionError
_________________________________________________________________ test_int_enum_successful_for_str_int _________________________________________________________________

    def test_int_enum_successful_for_str_int():
        m = CookingModel(tool='2')
        assert m.tool == ToolEnum.wrench
>       assert repr(m.tool) == '<ToolEnum.wrench: 2>'
E       AssertionError: assert 'ToolEnum.wrench' == '<ToolEnum.wrench: 2>'
E         - <ToolEnum.wrench: 2>
E         ? -               ----
E         + ToolEnum.wrench

tests/test_types.py:856: AssertionError

```

### PR DESCRIPTION
**Pull request by PrettyWood**
## Change Summary
Python 3.10.0b2 is out. Maintainers are invited to try it out. I reckon adding it in the CI doesn't hurt!

ℹ️ I invite you to **read this commit by commit**, which should make the review trivial

## Related issue number

closes #2607
closes #2786
closes #2787

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by PrettyWood** [comment]
Hey @samuelcolvin 
Since 3.10 was possibly a game breaker for _pydantic_, I reckon it's a good idea to check everything works as expected as soon as possible.
Quoting [the python release](https://www.python.org/downloads/release/python-3100b2/) itself 
> We strongly encourage maintainers of third-party Python projects to test with 3.10 during the beta phase

Reading this PR commit by commit should be trivial. The main thing is that I extracted `BaseConfig` and `Extra` in a dedicated module to make checks on enum values clean (only import this module instead of the whole `main` one)
Please review and tell me what you think 🙇

**Review by hmvp — APPROVED** [review]

**Review by samuelcolvin — COMMENTED** [review]
This looks great, pretty much ready to merge.

Can you just confirm what's changed in `config.py` vs. `main.py`?

File: pydantic/config.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+__all__ = 'BaseConfig', 'Extra', 'inherit_config', 'prepare_config'
+
+
+class Extra(str, Enum):
+    allow = 'allow'
+    ignore = 'ignore'
+    forbid = 'forbid'
+
+
+class BaseConfig:
```
[the remark this answers is not in the document]
Comment (PrettyWood) [reply pydantic/config.py:40]: Yes in the first commit 6a243aec6838e45930230cbffed9faaf870cb529 I just moved the code except `EXTRA_LINK = 'https://pydantic-docs.helpmanual.io/usage/model_config/'` that got removed since it was not used anymore

[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/config.py:40]: great.

**Comment by samuelcolvin** [comment]
Thanks so much.

Sorry for the slow responses. New baby, new job, no sleep, no open source for a little while.

**Comment by tirkarthi** [comment]
Just to add enum repr related changes were reverted in Python 3.10 after discussion : https://mail.python.org/archives/list/python-dev@python.org/thread/ZMC67QA2JVQJSWSFWRS6IM6ZX4EK277G/#LSTMFAPSPD3BGZ4D6HQFODXZVB3PLYKF

**Comment by PrettyWood** [comment]
Congrats @samuelcolvin for the baby I didnt' know!
And thanks @tirkarthi for the info.
I'll open another PR later when next python3.10 beta version is out and revert https://github.com/samuelcolvin/pydantic/pull/2885/commits/b1c8d9ef1396959ff9d88bb2ed16d99dd3146151.
The rest should not be impacted

**Comment by rmzr7** [comment]
@samuelcolvin when will this be released?

**Comment by m10d** [comment]
commenting here looking for advice from those familiar with the py3.10 changes.  We recently had to upgrade from `python3.8`->`3.10`.  However, we encountered a pydantic bug (#3896) that blocks us from upgrading `pydantic` to `1.9.0` - which I understand per the title here is the only version to officially support `python3.10`.  remaining on `pydantic==1.7.4` is the only workaround I am aware of.  A comment from samuel on that bug indicates its not trivial / won't likely be resolved very soon. 

We use pydantic for all serialization & deserialization for `json` input to and output from our api (and offline for schema & tool generation).  While our automated testing IS somewhat basic, both our `unit` and `api`-level full system tests are passing with `python3.10` + `pydantic==1.7.4`.  

*my question is* - any advice on common edge cases we should be aware of or add testing for, to (attempt to) verify this pair of versions will not bite us?  I've read the linked issues and I don't believe they affect our specific use of `pydantic`, but `3.10` has some significant changes from `3.8`.  Any advise or landmines would be greatly appreciated.

**Comment by samuelcolvin** [comment]
Best way to speedup a resolution and the release of 1.9.1 world be if you could submit a bugfix for #3896.

Otherwise I'll look at it soon.

**Comment by m10d** [comment]
> could submit a bugfix for https://github.com/samuelcolvin/pydantic/issues/3896.

You may have struck some fear into my heart with your comment on #3896 :) 

> "bug deep in schema.py I doubt there's an easy fix I'm afraid." -SC

If you take an initial look, and it's something suited for a new contributor, I might be able to dig in.  A deep dive is not in the cards for me at this movement, unfortunately.  Thank you for your replies, pydantic has long been a very helpful library.
